geo-traits/src/rect.rs

```diff
@@ -0,0 +1,85 @@
+use std::marker::PhantomData;
+
+use geo_types::{Coord, CoordNum, Rect};
+
+use crate::{Dimensions, PointTrait, UnimplementedPoint};
+
+/// A trait for accessing data from a generic Rect.
+///
+/// A Rect is an _axis-aligned_ bounded 2D rectangle whose area is
+/// defined by minimum and maximum [`Point`s][PointTrait].
+pub trait RectTrait {
+    /// The coordinate type of this geometry
+    type T: CoordNum;
+
+    /// The type of each underlying coordinate, which implements [PointTrait]
+    type PointType<'a>: 'a + PointTrait<T = Self::T>
+    where
+        Self: 'a;
+
+    /// The dimension of this geometry
+    fn dim(&self) -> Dimensions;
+
+    /// The minimum coordinate of this Rect
+    fn min(&self) -> Self::PointType<'_>;
+
+    /// The maximum coordinate of this Rect
+    fn max(&self) -> Self::PointType<'_>;
+}
```

Comment:
Similar to `Triangle`, what about a `points` method here? Akin to:

```rust
fn points() -> [Self::PointType; 4] {
    let min = self.min();
    let max = self.max();
    iter::once(min.x, max.x).chain(...).chain(...).chain(...)
}
```

I don't feel strongly though. I needed something like this for a project but it's no biggie

Reply:
Ah nevermind this would involve constructing new points, which wouldn't work I don't think

Reply:
One thing I've felt rather strongly about is that the traits should be read-only, constant-time, and not need to know how to create _new geometry objects_. Here, this `points` method would need to create new objects for each corner (and potentially to allocate as well). This `iter::once(min.x, max.x)` also seems like it would need to be generalized across all variants of `Dimensions`, right?

Reply:
Yep yep, what you said makes sense